### ISSUE TITLE
are there anyways to extend types in our own adapters?

### ISSUE THREAD

**Issue by shekhei** [balderdashy/waterline#819]
There are cases where certain types in the adapter are not quite available in the types list, and they all end up getting changed to string, which, i reckon, not quite desirable.

If there isnt, I think there should at least be a way to extend in your own adapter?

**Comment by dmarcelino** [comment]
Another related suggestion brought by @GaryLCoxJr on #665:

> I understand the waterline way which is to ensure all db adapters work.  Why don't we add support for the types a db uses, for example on Postgre uuid and decimal, and then for other adapters that dont use those types we just set a type it does use.  This way we can utilize the correct data types for a given database.
> 
> Example:
> Mongo doesn't have a uuid type, so that adapter might return a string
> ```javascript
> case "uuid":
>   return "string";
> ```
> 
> whereas on the postgre it would use that type.  This would keep things working the way waterline is designed and would allow us to use the correct data types.  I'm willing to add these types and update the existing adapters.
> 
> Update: Or we can just have a default catch all that returns a string if none of the case types matched.

**Comment by devinivy** [comment]
This can generally be dealt with in-adapter for special cases like ids (for example, in sails-dynamodb the primary attribute defaults to a string type that is auto-filled by a UUID).  Otherwise, often is the case that a waterline type plus a validator will do the trick.  It might not be an ideal solution, but these things work.  Consider writing mixins for adapter-specific validations.

**Comment by dmarcelino** [comment]
@shekhei came up with an interesting solution for this on PR #832, I'm not sure it's **the** solution but it's definitely an interesting approach. I would like to have a more *formal* way of defining both types and validations on adapters, I believe it's something that makes sense.

**Comment by sailsbot** [comment]
Thanks for posting, @shekhei.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
Reopening as this one has PRs #981 and balderdashy/anchor#77 associated with it.

**Comment by sailsbot** [comment]
Thanks for posting, @shekhei.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @shekhei.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
